File: curbcheck/geocode/intersection.py
"""The corner rung: a node both named streets meet at.

The pairs are pre-written by the ETL, so a query that names two streets is two
street resolutions and one indexed lookup per pair, never a graph walk.
"""

from __future__ import annotations

import sqlite3

from curbcheck.engine.labels import single_spaced
from curbcheck.geocode.candidates import (
    CONFIDENCE_INTERSECTION,
    FUZZY_CONFIDENCE_PENALTY,
    GeocodeCandidate,
    GeocodeKind,
)
from curbcheck.geocode.query import IntersectionQuery
from curbcheck.geocode.street import MAX_STREETS_PER_SIDE, resolve_street

_INTERSECTION_SQL = (
    "SELECT display, lon, lat FROM intersection WHERE a_norm = ? AND b_norm = ? LIMIT ?"
)


def intersection_candidates(
    conn: sqlite3.Connection, query: IntersectionQuery, limit: int
) -> list[GeocodeCandidate]:
    """Nodes where both streets meet, from the pre-paired `intersection` table."""
    firsts = resolve_street(conn, query.first)[:MAX_STREETS_PER_SIDE]
    seconds = resolve_street(conn, query.second)[:MAX_STREETS_PER_SIDE]
    found = []
    for a_norm, a_fuzzy in firsts:
        for b_norm, b_fuzzy in seconds:
            penalty = FUZZY_CONFIDENCE_PENALTY if (a_fuzzy or b_fuzzy) else 1.0
            for display, lon, lat in conn.execute(_INTERSECTION_SQL, (a_norm, b_norm, limit)):
                found.append(
                    GeocodeCandidate(
                        label=single_spaced(str(display)),
                        lat=float(lat),
                        lon=float(lon),
                        kind=GeocodeKind.INTERSECTION,
                        confidence=CONFIDENCE_INTERSECTION * penalty,
                    )
                )
    return found

File: site/static/geocode/intersection.js
/**
 * The corner rung: a node both named streets meet at.
 *
 * The pairs are pre-written by the ETL, so a query that names two streets is two
 * street resolutions and one indexed lookup per pair, never a graph walk.
 */

import { singleSpaced } from "../engine/labels.js";
import {
  CONFIDENCE_INTERSECTION,
  FUZZY_CONFIDENCE_PENALTY,
  GeocodeKind,
  geocodeCandidate,
} from "./candidates.js";
import { intersectionRange } from "./index.js";
import { MAX_STREETS_PER_SIDE, resolveStreet } from "./street.js";

/**
 * Nodes where both streets meet, from the pre-paired `intersection` table.
 *
 * Replaces `_INTERSECTION_SQL` (`… WHERE a_norm = ? AND b_norm = ? LIMIT ?`),
 * whose plan is `SEARCH intersection USING COVERING INDEX ix_intersection_pair`
 * with no `ORDER BY`: the `LIMIT` therefore takes the first rows of the index's
 * own `(lon, lat, display, rowid)` order within the pair, which
 * `index.intersectionOrder` holds.
 */
export function intersectionCandidates(pack, index, query, limit) {
  const table = pack.geocode.intersections;
  const firsts = resolveStreet(pack, index, query.first).slice(0, MAX_STREETS_PER_SIDE);
  const seconds = resolveStreet(pack, index, query.second).slice(0, MAX_STREETS_PER_SIDE);
  const found = [];
  for (const [aNorm, aFuzzy] of firsts) {
    for (const [bNorm, bFuzzy] of seconds) {
      const penalty = aFuzzy || bFuzzy ? FUZZY_CONFIDENCE_PENALTY : 1.0;
      const [start, end] = intersectionRange(pack, index, aNorm, bNorm);
      for (let i = start; i < end && i - start < limit; i += 1) {
        const row = index.intersectionOrder[i];
        found.push(
          geocodeCandidate({
            label: singleSpaced(table.display[row]),
            lat: table.lat[row],
            lon: table.lon[row],
            kind: GeocodeKind.INTERSECTION,
            confidence: CONFIDENCE_INTERSECTION * penalty,
          }),
        );
      }
    }
  }
  return found;
}
